core/storage/translation/gae_models.py

```diff
@@ -0,0 +1,147 @@
+# coding: utf-8
+#
+# Copyright 2021 The Oppia Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Models for machine translation."""
+
+from __future__ import absolute_import  # pylint: disable=import-only-modules
+from __future__ import unicode_literals  # pylint: disable=import-only-modules
+
+from core.platform import models
+import utils
+
+(base_models,) = models.Registry.import_models(
+    [models.NAMES.base_model])
+
+datastore_services = models.Registry.import_datastore_services()
+
+
+class MachineTranslatedTextModel(base_models.BaseModel):
+    """Model for storing machine translations. Model instances have a key
+    generated from the source and target language codes, followed by a sha-1
+    hash of the origin text formated as follows:
+
+        [source_language_code]:[target_language_code]:[hashed_origin_text]
+
+    See MachineTranslatedTextModel._generate_id() below for details."""
+
+    # The untranslated text.
+    origin_text = datastore_services.TextProperty(required=True, indexed=False)
+    # An sha-1 hash of the origin text.
+    hashed_origin_text = datastore_services.StringProperty(
+        required=True, indexed=True)
+    # The language code of the untranslated text.
+    source_language_code = datastore_services.StringProperty(
+        required=True, indexed=True)
+    # The language code of the translation.
+    target_language_code = datastore_services.StringProperty(
+        required=True, indexed=True)
+    # The translation.
+    translated_text = datastore_services.TextProperty(
+        required=True, indexed=False)
+
+    @classmethod
+    def create(
+            cls, source_language_code, target_language_code, origin_text,
+            translated_text):
+        """Creates a new MachineTranslatedTextModel instance and returns its
+        ID.
+
+        Args:
+            source_language_code: str. The language code of the untranslated
+                text.
+            target_language_code: str. The language code of the translation.
+            origin_text: str. The untranslated text.
+            translated_text: str. The translation.
+
+        Returns:
+            str. The id of the newly created MachineTranslatedTextInstance.
+        """
+        hashed_origin_text = utils.convert_to_hash(origin_text, 50)
```

Comment:
Not related to  your PR, but @seanlip why do we have a length limit for convert_to_hash? Unless I'm mistaken, SHA-1 always produces a 160 bit string.

Further, we might want to consider being explicit about the hash name. If storage models & other APIs specifically need to depend on SHA-1, I'd expect them to call something like ``utils.convert_to_sha1_hash`` or similar to ensure their contracts can be upheld.

Reply:
I added a note here clarifying the choice of max_length.